app/src/main/java/pm/gnosis/heimdall/ui/safe/recover/extension/ReplaceExtensionStartActivity.kt

```diff
@@ -0,0 +1,211 @@
+                feesError.visibility = View.VISIBLE
+            }
+            progressBar.visibility = View.GONE
+            balanceBeforeValue.text = it.balanceBefore
+            feesValue.text = it.fee.underline()
+            balanceAfterValue.text = it.balanceAfter
+
+        })
+
+    }
+
+    override fun onStart() {
+        super.onStart()
+
+        backArrow.setOnClickListener {
+            finish()
+        }
+
+        feesValue.setOnClickListener {
+            startActivity(PaymentTokensActivity.createIntent(this, safe))
+        }
+
+        disposables += feesInfo.clicks()
+            .subscribeBy {
+                InfoTipDialogBuilder.build(this, R.layout.dialog_network_fee, R.string.ok).show()
+            }
+
+        disposables += bottomPanel.forwardClicks.subscribeBy {
+            startActivity(ReplaceExtensionQrActivity.createIntent(this, safe))
+        }
+
+        viewModel.estimate()
+    }
+
+
+    companion object {
+
+        private const val EXTRA_SAFE_ADDRESS = "extra.string.safe_address"
+
+        fun createIntent(context: Context, safeAddress: Solidity.Address) = Intent(context, ReplaceExtensionStartActivity::class.java).apply {
+            putExtra(EXTRA_SAFE_ADDRESS, safeAddress.asEthereumAddressString())
+        }
+    }
+}
+
+
+abstract class ReplaceExtensionStartContract : ViewModel() {
+
+    abstract val observableState: LiveData<ViewUpdate>
+
+    abstract fun setup(safeAddress: Solidity.Address)
+
+    abstract fun estimate()
+
+
+    data class ViewUpdate(
+        val balanceBefore: String,
+        val fee: String,
+        val balanceAfter: String,
+        val tokenSymbol: String,
+        val sufficient: Boolean
+    )
+}
+
+
+class ReplaceExtensionStartViewModel @Inject constructor(
+    private val recoverSafeOwnersHelper: RecoverSafeOwnersHelper,
+    private val gnosisSafeRepository: GnosisSafeRepository,
+    private val tokenRepository: TokenRepository,
+    private val transactionExecutionRepository: TransactionExecutionRepository
+
+) : ReplaceExtensionStartContract() {
+
+    override val observableState: LiveData<ViewUpdate>
+        get() = _state
+    private val _state = MutableLiveData<ViewUpdate>()
+
+    private lateinit var safeAddress: Solidity.Address
+
+    override fun setup(safeAddress: Solidity.Address) {
+        this.safeAddress = safeAddress
+    }
+
+    override fun estimate() {
+
+        viewModelScope.launch {
+
+            val safeInfo = gnosisSafeRepository.loadInfo(safeAddress).awaitFirst()
+            val paymentToken = tokenRepository.loadPaymentToken(safeAddress).await()
+
+
+            val balance = tokenRepository.loadTokenBalances(safeAddress, listOf(paymentToken))
+                .repeatWhen { it.delay(BALANCE_REQUEST_INTERVAL_SECONDS, TimeUnit.SECONDS) }
+                .retryWhen { it.delay(BALANCE_REQUEST_INTERVAL_SECONDS, TimeUnit.SECONDS) }
+                .awaitFirst()[0]
```

Comment:
If we `awaitFirst()` this is not really repeating, right? The UI should keep updating over time (not only once).

Reply:
should we also update gas fee value using same update interval?

Reply:
hmmm maybe that is a good idea ... even so it might alos be confusing. But it is more correct ... so I would vote for yes. @tschubotz any preferences?